planning/MartenV4.md

```diff
@@ -0,0 +1,405 @@
+        // get at what the identity of the aggregate document is
+        // Prefer the FI version of this though
+        public int AggregateBy(Event1 @event);
+        public int AggregateBy(EventWrapper @event);
+
+        // Sync version with raw event
+        public void Apply(Event1 @event, ProjectedDocument doc);
+
+        // Sync version with event wrapper, could be a custom implementation of IEvent
+        // to get at event metadata & stream metadata or partitioning information
+        public void Apply(EventWrapper<Event1> @event, MyEventStream stream, ProjectedDocument doc);
+
+        // Sync version w/ immutable projected docs
+        public ProjectedDocument Apply(Event1 @event, ProjectedDocument doc);
+
+        // apply a partial update to an aggregate document, but this wouldn't
+        // get used if there's some reason to fetch the whole aggregate for a segment
+        // of events. 
+        public void Partial(Event1, @event, IDocumentSession session);
+
+        // Should the aggregate document be deleted based on some event data?
+        public bool ShouldBeDeleted(Event1 @event);
+        public bool ShouldBeDeleted(EventWrapper @event);
+        public bool ShouldBeDeleted(EventWrapper @event, MyEventStream stream);
+    }
+```
+
+#### Defining a Projection to a SQL Statement(s)
+
+TODO
+
+#### Inline Projections
+
+There's been some complaints and obvious confusion about the `IEvent` interface and how the metadata properties like
+`version` aren't available during inline projections. That's because today the inline projections are all calculated 
+just prior to submitting the database commands before the version numbers of each event within the stream would be calculated. 
+
+In v4, the inline projection support and the `UpdateBatch` mechanics should be smart enough to "know" when the transformation
+or aggregation requires metadata, so that:
+
+#### If no event or stream metadata is necessary...
+
+It works basically as it does today. Upon `IDocumentSession.SaveChanges() / SaveChangesAsync()`:
+
+1. The inline projections work across the incoming event objects, create new projected documents
+or update existing documents as necessary just prior to sending the current `UnitOfWork` to the database
+1. The `DocumentSession` issues the batched up commands with both the event capture and the projection documents
+1. The underlying database transaction is committed
+
+If there's no need for event metadata, do thing in fewer network round trips
+
+#### If event or stream metadata is necessary...
+
+This gets a little more complicated, because the projections can't be calculated until *after* the events are appended, so:
+
+1. The events get appended, and the sproc returns an array of the new event versions
+1. Build up an array of `IEvent` wrappers that have the stream and event metadata
+1. Pass the `IEvent` wrappers into the transformation code, and get an additional `UpdateBatch` of projected document inserts
```

Comment:
As I wrote in the previous comment. It would be good if we make the convention mapping of the metadata to the event so we would not force users to implement `IEvent` as in my opinion it's making things much more complicated and less intuitive.

Reply:
You'd only need a custom `IEvent` type if you wanted custom metadata. I'm not super thrilled about having to have a Dictionary for metadata. I could be convinced I suppose.

Reply:
I'm also not in favour of Dictionary. Maybe we could come up with some convention mapping (so eg. if event has field Timestamp then it might inherit the Timestamp from the metadata).